### PR TITLE
Initial work with swarm and ETCD as a discovery service

### PR DESCRIPTION
**Pull request by chuckbutler**
this is NOT READY TO MERGE

- Launches a swarm cluster when related to ETCD
- Launches a manager node on the current leader

still TODO:

- Generate TLS certificates and secure the docker daemon behind TLS
  connections
- An action to generate a docker config bundle for use with a docker
  workstation (config, tls certs, ca cert)

To view the ETCD relation stub -
https://github.com/chuckbutler/interface-etcd.git

@mbruzek @bcsaller If you get some free time, could you give this a once over? 

I have a few thoughts about how I've done this, and I'll do my best to compose them below:

This layer might be better off as a secondary "mixin" layer - which you include implicitly in your `composer.yaml`. As it's not always the end-goal to provide SWARM with your charm. And in fact, it might complicate things (Like with K8's - this layer has a goal of replacing the existing docker charm when it matures)

Should swarm be a configuration option to enable/disable the entirety of the swarm components?
if not swarm_requested():
     return
Plastering that around seemed iffy, which is why i think the method above might be smarter, but again, it becomes a component more than a "feature".

When standing up, relating, unrelating, and tearing down - it seems like its possible to make the cluster inconsistent.

To validate what you've stood up, you can do the following from any node that has reachability to your swarm cluster (including your laptop)

    docker -H tcp://<ip of swarm manager>:2377 info

you should get output similar to the following:

```
Containers: 6
Images: 4
Role: primary
Strategy: spread
Filters: affinity, health, constraint, port, dependency
Nodes: 2
 amused-brass: 192.168.100.50:2375
  └ Containers: 3
  └ Reserved CPUs: 0 / 1
  └ Reserved Memory: 0 B / 2.053 GiB
  └ Labels: executiondriver=native-0.2, kernelversion=3.13.0-63-generic, operatingsystem=Ubuntu 14.04.3 LTS, storagedriver=aufs
 joyous-cap: 192.168.100.53:2375
  └ Containers: 3
  └ Reserved CPUs: 0 / 1
  └ Reserved Memory: 0 B / 2.053 GiB
  └ Labels: executiondriver=native-0.2, kernelversion=3.13.0-63-generic, operatingsystem=Ubuntu 14.04.3 LTS, storagedriver=aufs
CPUs: 2
Total Memory: 4.106 GiB
Name: d09c853a5040
```

Additionally, you can launch containers on the docker port 2375

    docker -H tcp://192.168.100.53:2375 run -d lazypower/github-high-scores

Targeting any unit of the cluster should be "intercepted" by the manager, and launched across a cluster member host who has the capacity to run it. **untested**  This needs more validation, and an amulet test to verify

The bundle used to launch the test cluster looks similar to the following:
```
services:
  docker:
    charm: local:trusty/docker
    num_units: 2
  etcd:
    charm: cs:trusty/etcd
relations:
  - ["docker", "etcd"]
series: trusty
```

If you see anything heinous in the code, just let me know. Thanks in advance to anyone who takes a look!

### CODE REVIEW

File: reactive/swarm.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    # TODO: refactor to be process run
+    cmd = "docker run -d --name swarmagent swarm join --advertise={0}:{1} {2}/swarm".format(addr, 2375, connection_string)  # noqa
+    check_call(split(cmd))
+    hookenv.open_port(2375)
+
+
+def start_swarm_etcd_manager(connection_string):
+    hookenv.status_set('maintenance', 'Starting swarm manager')
+    # TODO: refactor to be process run
+    cmd = "docker run -d --name swarmmanager -p 2377:2375 swarm manage {}/swarm".format(connection_string)  # noqa
```
Comment (mbruzek) [inline reactive/swarm.py]: 2377 -> 2375 ?  Why do you map 2377 to 2375?

File: reactive/swarm.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+@when_not('swarm.available')
+def swarm_etcd_cluster_setup(etcd):
+    """
+    Expose the Docker TCP port, and begin swarm cluster configuration. Always
+    leading with the agent, connecting to the discovery service, then follow
+    up with the manager container on the leader node.
+    """
+    bind_docker_daemon()
+    start_swarm_etcd_agent(etcd.connection_string())
+    if hookenv.is_leader():
```
Comment (mbruzek) [inline reactive/swarm.py]: Only the leader needs the etcd manager?  Do the other charms need to know etcd also?

**Comment by mbruzek** [comment]
Hey Chuck,

I had some time to take a look today and it looks reasonable.  I had a few questions in-line but no blockers. 

I suspect you have iterated on this concept since or preparing for the demo/lightning talk.  Has the concept solidified with use or has it changed in any significant way?

File: reactive/swarm.py (hunk 1 above)
Comment (chuckbutler) [inline reactive/swarm.py]: 2375 is occupied by the docker daemon. This maps 2377 on the host to 2375 in the container. This is what enables the swarm manager to act as a docker daemon proxy to do placement et-al. when you call it like so:

docker -H tcp://<ip>:2377 info

File: reactive/swarm.py (hunk 2 above)
Comment (chuckbutler) [inline reactive/swarm.py]: The swarm agents connect to the swarm manager, and the swarm manager communicates to ETCD on behalf of the cluster.

**Comment by chuckbutler** [comment]
Thanks for taking a look @mbruzek. I'm still working through the logistics of the SSL/TLS termination of the docker daemon. Until we have that bit, I'm not confident in merging this for consumption. Running an open Docker Daemon is potentially hazardous.

**Pull request closed without merging**